### ISSUE TITLE
I can't run downscaling example on the cloud through Lambda function

### ISSUE THREAD

**Issue by mpu-creare** [creare-com/podpac#202]
**Description**
/podpac-examples/notebooks/demos/SMAP-downscaling-example-application.ipynb

**Steps to Reproduce**
```
l = podpac.core.managers.aws_lambda.Lambda(source=downscaled_sm)
l.eval(c)
```
Then traceback on cloudwatch:
```
Could not parse coordinates definition item with keys odict_keys(['name', 'coord_ref_sys', 'ctype']): ValueError
Traceback (most recent call last):
File "/var/task/handler.py", line 52, in handler
pipeline.eval(coords)
File "/tmp/podpac/core/pipeline/pipeline.py", line 102, in eval
output = self.output.node.eval(coordinates, output)
File "/var/task/traitlets/traitlets.py", line 556, in __get__
return self.get(obj, cls)
File "/var/task/traitlets/traitlets.py", line 535, in get
value = self._validate(obj, dynamic_default())
File "/tmp/podpac/core/pipeline/pipeline.py", line 89, in _parse_definition
return parse_pipeline_definition(self.definition)
File "/tmp/podpac/core/pipeline/pipeline.py", line 147, in parse_pipeline_definition
nodes[key] = _parse_node_definition(nodes, key, d)
File "/tmp/podpac/core/pipeline/pipeline.py", line 222, in _parse_node_definition
return node_class(**kwargs)
File "/tmp/podpac/core/node.py", line 158, in __init__
tkwargs = self._first_init(**kwargs)
File "/tmp/podpac/core/data/types.py", line 1032, in _first_init
kwargs['reprojected_coordinates'] = Coordinates.from_definition(kwargs['reprojected_coordinates'])
File "/tmp/podpac/core/coordinates/coordinates.py", line 346, in from_definition
raise ValueError("Could not parse coordinates definition item with keys %s" % elem.keys())
ValueError: Could not parse coordinates definition item with keys odict_keys(['name', 'coord_ref_sys', 'ctype'])
```

**Expected Behavior**
This should work

**Observed Behavior**
This doesn't work. 

**Additional Notes**
I'm surprised that it doesn't work. When I try to reproduce what the lambda function should do locally, it works just fine. I can easily recreate the pipeline.

**Comment by jmilloy** [comment]
The issue isn't serializing/deserializing the node, but serializing/deserialising the coordinates. Can you include creating `c` in your steps to reproduce? We should look at the output of `c.definition` and try running `Coordinates.from_definition(c.definition)` locally to try and reproduce.

**Comment by mpu-creare** [comment]
no rush.

**Comment by mpu-creare** [comment]
I can't reproduce locally. We'll make a plan next week.

**Comment by dsully1992** [comment]
@mpu-creare Can you coordinates in the steps to reproduce? We can debug the handler to pinpoint whether the coordinates or pipeline are being serialized or deserialized incorrectly

**Comment by jmilloy** [comment]
It's the coordinates, the coordinates object it's trying to deserialize has no list of values or start/stop/step. It only has name, coord_ref_sys, and ctype, but it needs either values or start/stop/step or start/stop/size.

**Comment by mpu-creare** [comment]
I'm trying to run this example: https://github.com/creare-com/podpac-examples/blob/develop/notebooks/demos/SMAP-downscaling-example-application.ipynb 

The coordinates that give the problems, IIRC, is from this part of the pipeline, where the `reprojected_coordinates` are not being correctly deserialized on the handler:

```python
twi_bar = podpac.data.ReprojectedSource(source=twi, 
                                        reprojected_coordinates=smap.shared_coordinates,       
                                        interpolation='bilinear', cache_type='disk')
```

It's worth trying to reproduce locally, but I wasn't able to do so. I'm guessing the podpac version in the lambda function is just a little out of date.

**Comment by mpu-creare** [comment]
Well... either that or something in the json serialization/deserialization that's doing something funny.

**Comment by dsully1992** [comment]
I see. Let me know if you'd like an updated AWS Lambda function.

**Comment by mpu-creare** [comment]
I think testing the handler in the docker image locally would be worth doing before updating the AWS Lambda... I also think it's worth automating the process at this point.

**Comment by dsully1992** [comment]
I'll sort it out locally in the docker, with a little massaging to mock using the handler.

Let's talk more, separately about automating the process.

**Comment by dsully1992** [comment]
In only briefly attempting this in the Docker container, here is where I run into problems:

```python
In [7]: twi_bar = podpac.data.ReprojectedSource(source=twi,  
   ...:                                         reprojected_coordinates=smap.shared_coord
   ...: inates,        
   ...:                                         interpolation='bilinear', cache_type='dis
   ...: k')                                                                              
/tmp/vendored/podpac/datalib/smap.py:150: UserWarning: WARNING: cannot authenticate to https://n5eil02u.ecs.nsidc.org/opendap/SMAP:HTTP error: <404>
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /SMAP was not found on this server.</p>
</body></html>

  warnings.warn('WARNING: cannot authenticate to {}:'.format(url) + str(e))
/tmp/vendored/podpac/datalib/smap.py:150: UserWarning: WARNING: cannot authenticate to https://n5eil02u.ecs.nsidc.org/opendap/SMAP/SPL4SMAU.004:HTTP error: <404>
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /SMAP/SPL4SMAU.004 was not found on this server.</p>
</body></html>

  warnings.warn('WARNING: cannot authenticate to {}:'.format(url) + str(e))
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    527         try:
--> 528             value = obj._trait_values[self.name]
    529         except KeyError:

KeyError: 'shared_coordinates'

During handling of the above exception, another exception occurred:

NodeException                             Traceback (most recent call last)
/tmp/vendored/podpac/core/node.py in cached_function()
    669                 try:
--> 670                     out = self.get_cache(key)
    671                 except NodeException:

/tmp/vendored/podpac/core/node.py in get_cache(self, key, coordinates)
    455         if not self.has_cache(key, coordinates=coordinates):
--> 456             raise NodeException("cached data not found for key '%s' and cooordinates %s" % (key, coordinates))
    457 

NodeException: cached data not found for key 'shared.coordinates' and cooordinates None

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
<ipython-input-7-9a87960b821a> in <module>
      1 twi_bar = podpac.data.ReprojectedSource(source=twi, 
----> 2                                         reprojected_coordinates=smap.shared_coordinates,
      3                                         interpolation='bilinear', cache_type='disk')

/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in __get__(self, obj, cls)
    554             return self
    555         else:
--> 556             return self.get(obj, cls)
    557 
    558     def set(self, obj, value):

/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    533                 raise TraitError("No default value found for %s trait of %r"
    534                                  % (self.name, obj))
--> 535             value = self._validate(obj, dynamic_default())
    536             obj._trait_values[self.name] = value
    537             return value

/tmp/vendored/podpac/core/compositor.py in _shared_coordinates_default(self)
    107     @tl.default('shared_coordinates')
    108     def _shared_coordinates_default(self):
--> 109         return self.get_shared_coordinates()
    110 
    111     def get_shared_coordinates(self):

/tmp/vendored/podpac/core/node.py in cache_wrapper(self)
    680 
    681             # Return the value on the first run
--> 682             return cached_function()
    683         return cache_wrapper
    684     return cache_decorator

/tmp/vendored/podpac/core/node.py in cached_function()
    670                     out = self.get_cache(key)
    671                 except NodeException:
--> 672                     out = func(self)
    673                     self.put_cache(out, key, raise_no_cache_exception=raise_no_cache_exception)
    674                 return out

/tmp/vendored/podpac/datalib/smap.py in get_shared_coordinates(self)
    904 
    905         coords = SMAPDateFolder(product=self.product, version=self.version,
--> 906                                 folder_date=self.get_available_times_dates()[1][0],
    907                                 auth_session=self.auth_session,
    908                                ).shared_coordinates

IndexError: list index out of range
```

Let's discuss at our meeting today.

**Comment by dsully1992** [comment]
```python
twi_bar = podpac.data.ReprojectedSource(source=twi,  
    ...:                                         reprojected_coordinates=smap.shared_coor
    ...: dinates,        
    ...:                                         interpolation='bilinear', cache_type='di
    ...: sk')                                                                            
/tmp/vendored/podpac/datalib/smap.py:150: UserWarning: WARNING: cannot authenticate to https://n5eil02u.ecs.nsidc.org/opendap/SMAP/SPL4SMAU.004:HTTP error: <404>
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /SMAP/SPL4SMAU.004 was not found on this server.</p>
</body></html>

  warnings.warn('WARNING: cannot authenticate to {}:'.format(url) + str(e))
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    527         try:
--> 528             value = obj._trait_values[self.name]
    529         except KeyError:

KeyError: 'shared_coordinates'

During handling of the above exception, another exception occurred:

NodeException                             Traceback (most recent call last)
/tmp/vendored/podpac/core/node.py in cached_function()
    669                 try:
--> 670                     out = self.get_cache(key)
    671                 except NodeException:

/tmp/vendored/podpac/core/node.py in get_cache(self, key, coordinates)
    455         if not self.has_cache(key, coordinates=coordinates):
--> 456             raise NodeException("cached data not found for key '%s' and cooordinates %s" % (key, coordinates))
    457 

NodeException: cached data not found for key 'shared.coordinates' and cooordinates None

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
<ipython-input-25-9a87960b821a> in <module>
      1 twi_bar = podpac.data.ReprojectedSource(source=twi, 
----> 2                                         reprojected_coordinates=smap.shared_coordinates,
      3                                         interpolation='bilinear', cache_type='disk')

/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in __get__(self, obj, cls)
    554             return self
    555         else:
--> 556             return self.get(obj, cls)
    557 
    558     def set(self, obj, value):

/usr/local/lib/python3.6/site-packages/traitlets/traitlets.py in get(self, obj, cls)
    533                 raise TraitError("No default value found for %s trait of %r"
    534                                  % (self.name, obj))
--> 535             value = self._validate(obj, dynamic_default())
    536             obj._trait_values[self.name] = value
    537             return value

/tmp/vendored/podpac/core/compositor.py in _shared_coordinates_default(self)
    107     @tl.default('shared_coordinates')
    108     def _shared_coordinates_default(self):
--> 109         return self.get_shared_coordinates()
    110 
    111     def get_shared_coordinates(self):

/tmp/vendored/podpac/core/node.py in cached_function()
    670                     out = self.get_cache(key)
    671                 except NodeException:
--> 672                     out = func(self)
    673                     self.put_cache(out, key, raise_no_cache_exception=raise_no_cache_exception)
    674                 return out

/tmp/vendored/podpac/datalib/smap.py in get_shared_coordinates(self)
    904 
    905         coords = SMAPDateFolder(product=self.product, version=self.version,
--> 906                                 folder_date=self.get_available_times_dates()[1][0],
    907                                 auth_session=self.auth_session,
    908                                ).shared_coordinates

IndexError: list index out of range
```

**Comment by mpu-creare** [comment]
Ah well. that looks like an authentication error with a terrible traceback but an actually very appropriate error message at the head of it all..

**Comment by dsully1992** [comment]
Agreed. I'm leaning toward just updating the AWS Lambda to include our latest changes.

**Comment by dsully1992** [comment]
Note: I plan to work on #146 and hope to use the result of that to update the Lambda to resolve this issue.

**Comment by dsully1992** [comment]
I am working on updating the Lambda function, but can't get around this issue with lazy-import:

```bash
pip3 install lazy-import
Collecting lazy-import
  Downloading https://files.pythonhosted.org/packages/44/2e/5378f9b9cbc893826c2ecb022646c97ece9efbaad351adf89425fff33990/lazy_import-0.2.2.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-4gksvuzd/lazy-import/setup.py", line 6, in <module>
        readme = infile.read()
      File "/usr/local/lib/python3.6/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 621: ordinal not in range(128)
```

**Comment by jmilloy** [comment]
Weird. That's usually a python 2 vs 3 issue. I'm using python 3.5 and installed lazy-import 0.2.2 with pip (on ubuntu) last week just fine. I know that's not much help.

**Comment by mpu-creare** [comment]
That looks like an error happened, and the error contains bytes that cannot be decoded. 

Try installing a specific version of lazy-import. Also try installing an older version, e.g. 0.2.1. 

That is very strange...

**Comment by dsully1992** [comment]
Ok, I think I found a fix. I based the Dockerfile off other forum posts in which the authors installed python3 using curl and make, because the amazon linux image is so trimmed down. However, the yum repo in that image has python37 available for install. Lazy-import installs when I use pip3.7 from that yum installation, so I'll update the Docker to use yum's python/pip.

**Comment by dsully1992** [comment]
Update: I have spent far more time on this than anticipated. The docker container can run podpac and all of its dependencies, but the AWS Lambda function complains about a broken installation of numpy. I'm guessing this has to do with how the zips are put together. Here is the stacktrace: 
```bash
IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the multiarray numpy extension module failed. Most
likely you are trying to import a failed build of numpy.
Here is how to proceed:
- If you're working with a numpy git repository, try `git clean -xdf`
(removes all files not under version control) and rebuild numpy.
- If you are simply trying to use the numpy version that you have installed:
your installation is broken - please reinstall numpy.
- If you have already reinstalled and that did not fix the problem, then:
1. Check that you are using the Python you expect (you're using /var/lang/bin/python3.6),
and that you have no directories in your PATH or PYTHONPATH that can
interfere with the Python and numpy versions you're trying to use.
2. If (1) looks fine, you can open a new issue at
https://github.com/numpy/numpy/issues. Please include details on:
- how you installed Python
- how you installed numpy
- your operating system
- whether or not you have multiple versions of Python installed
- if you built from source, your compiler versions and ideally a build log

Note: this error has many possible causes, so please don't comment on
an existing issue about this - open a new one instead.

Original error was: No module named 'numpy.core._multiarray_umath'
: ImportError
Traceback (most recent call last):
File "/var/task/handler.py", line 45, in handler
import matplotlib
File "/tmp/matplotlib/__init__.py", line 141, in <module>
from . import cbook, rcsetup
File "/tmp/matplotlib/cbook/__init__.py", line 33, in <module>
import numpy as np
File "/var/task/numpy/__init__.py", line 142, in <module>
from . import core
File "/var/task/numpy/core/__init__.py", line 47, in <module>
raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the multiarray numpy extension module failed. Most
likely you are trying to import a failed build of numpy.
Here is how to proceed:
- If you're working with a numpy git repository, try `git clean -xdf`
(removes all files not under version control) and rebuild numpy.
- If you are simply trying to use the numpy version that you have installed:
your installation is broken - please reinstall numpy.
- If you have already reinstalled and that did not fix the problem, then:
1. Check that you are using the Python you expect (you're using /var/lang/bin/python3.6),
and that you have no directories in your PATH or PYTHONPATH that can
interfere with the Python and numpy versions you're trying to use.
2. If (1) looks fine, you can open a new issue at
https://github.com/numpy/numpy/issues. Please include details on:
- how you installed Python
- how you installed numpy
- your operating system
- whether or not you have multiple versions of Python installed
- if you built from source, your compiler versions and ideally a build log

Note: this error has many possible causes, so please don't comment on
an existing issue about this - open a new one instead.

Original error was: No module named 'numpy.core._multiarray_umath'
```

**Comment by dsully1992** [comment]
I spent a bit of time looking into this and attempting fixes. In my latest attempt, the docker mounts the local podpac directory (which I have installed on my machine), copies those settings, runs `git clean -xdf` to remove files added by the local installation of podpac, then installs podpac in a new directory designated for the archive files. I may revert to cloning a fresh podpac in the docker image, though this issue seems related to using yum's python 3.7 instead of the manually installed python 3.6.

**Comment by mpu-creare** [comment]
Let's try to get this done early in 0.4.0 -- this seems like an important issue that needs to be resolved carefully.

**Comment by dsully1992** [comment]
Ah right because I've unintentionally upgraded the docker image to python 3.7, that numpy installation is compiled for python 3.7, not 3.6. I'll try to find another fix to install lazy-import in the docker image with python 3.6.

**Comment by dsully1992** [comment]
Ok, so lazy-import's README.rst has non-ascii characters. I'm unclear on why some installations of python 3.6 (my windows, MacOS, and other local machines) encodings module can handle this when installing lazy-import. Lazy import has an issue and pull request to fix this; however, AWS Lambda functions now have a Python 3.7 environment. So, I've updated our Lambda functions to use that environment, which gets us around the numpy issue as well - now the Docker can install podpac with Python 3.7. 

That said, there is a new issue in the Lambda function, which seems like something I'm better equipped to resolve. Here is that stacktrace: 

```bash
[ERROR] ValueError: Could not parse coordinates definition item with keys odict_keys(['name', 'coord_ref_sys', 'ctype'])
Traceback (most recent call last):
  File "/var/task/handler.py", line 53, in handler
    pipeline.eval(coords)
  File "/tmp/podpac/core/pipeline/pipeline.py", line 102, in eval
    output = self.output.node.eval(coordinates, output)
  File "/var/task/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/var/task/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/tmp/podpac/core/pipeline/pipeline.py", line 89, in _parse_definition
    return parse_pipeline_definition(self.definition)
  File "/tmp/podpac/core/pipeline/pipeline.py", line 147, in parse_pipeline_definition
    nodes[key] = _parse_node_definition(nodes, key, d)
  File "/tmp/podpac/core/pipeline/pipeline.py", line 222, in _parse_node_definition
    return node_class(**kwargs)
  File "/tmp/podpac/core/node.py", line 155, in __init__
    tkwargs = self._first_init(**kwargs)
  File "/tmp/podpac/core/data/types.py", line 1040, in _first_init
    kwargs['reprojected_coordinates'] = Coordinates.from_definition(kwargs['reprojected_coordinates'])
  File "/tmp/podpac/core/coordinates/coordinates.py", line 346, in from_definition
    raise ValueError("Could not parse coordinates definition item with keys %s" % elem.keys())
```

@mpu-creare Does this look like the issue you've solved on the develop branch?

**Comment by mpu-creare** [comment]
What does the JSON look like for your Lambda function call?

**Comment by dsully1992** [comment]
I sent you an e-mail linking to it on S3.

**Comment by jmilloy** [comment]
Can I get that link, too?

**Comment by dsully1992** [comment]
Sure, I almost posted the JSON here, but it is large, and I didn't want to comb through it to ensure there is no private information.

**Comment by dsully1992** [comment]
Note, the JSON was created by following [this notebook](https://github.com/creare-com/podpac-examples/blob/develop/notebooks/demos/SMAP-downscaling-example-application.ipynb)

**Comment by jmilloy** [comment]
I don't think this will solve this problem, but that json is now out of date after the merge this morning, and you will have other issues. You should pull and regenerate the json.

**Comment by dsully1992** [comment]
I figured my branch is slightly out of date, since I know there have been other changes to develop since my last merge. The reason I was hoping to talk it out here before updating is I have to recreate the AWS Lambda function in parallel with any local pull, so it is running the same version of Podpac as I am locally. What is out of date about it? It seems likely that it being out of date means I should update the Lambda function, but I'm hoping to understand why first.

**Comment by jmilloy** [comment]
Ah, I semi-reproduced this. I'm on the hunt. I've got to be somewhere till about 1:30, but I'll get back on it this afternoon.

To reproduce locally, you can load the json (`d` below), and you can load the reprojected_coordinates from_definition the first time but then it removes the values key for some reason and you can't do it the second time. Should be something I can fix pretty quickly now that I've narrowed it down.

```
In[26]: podpac.Coordinates.from_definition(d['pipeline']['nodes']['TopographicWetnessIndexComposited3090m_reprojected']['attrs']['reprojected
    ...: _coordinates'])
Out[26]: 
Coordinates
	lat: ArrayCoordinates1d(lat): Bounds[-84.65644073486328, 84.65644073486328], N[1624], ctype['midpoint']
	lon: ArrayCoordinates1d(lon): Bounds[-179.95474243164062, 179.95474243164062], N[3856], ctype['midpoint']

In [27]: d['pipeline']['nodes']['TopographicWetnessIndexComposited3090m_reprojected']
Out[27]: 
OrderedDict([('interpolation', 'bilinear'),
             ('lookup_source', 'TopographicWetnessIndexComposited3090m'),
             ('attrs',
              OrderedDict([('reprojected_coordinates',
                            [OrderedDict([('ctype', 'midpoint'),
                                          ('name', 'lat'),
                                          ('coord_ref_sys', 'WGS84')]),
                             OrderedDict([('ctype', 'midpoint'),
                                          ('name', 'lon'),
                                          ('coord_ref_sys', 'WGS84')])]),
                           ('source_interpolation', 'nearest_preview')])),
             ('node', 'core.data.types.ReprojectedSource')])

In [28]: podpac.Coordinates.from_definition(d['pipeline']['nodes']['TopographicWetnessIndexComposited3090m_reprojected']['attrs']['reprojected
    ...: _coordinates'])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-1e18c49cf69c> in <module>()
----> 1 podpac.Coordinates.from_definition(d['pipeline']['nodes']['TopographicWetnessIndexComposited3090m_reprojected']['attrs']['reprojected_coordinates'])

~/Pipeline/podpac/podpac/core/coordinates/coordinates.py in from_definition(cls, d)
    368                 c = ArrayCoordinates1d.from_definition(elem)
    369             else:
--> 370                 raise ValueError("Could not parse coordinates definition item with keys %s" % elem.keys())
    371 
    372             coords.append(c)

ValueError: Could not parse coordinates definition item with keys odict_keys(['ctype', 'name', 'coord_ref_sys'])
```

**Comment by dsully1992** [comment]
Thanks, Jeff. I'll hold off for now, then pull develop, merge it into my feature branch, and update the AWS Lambda function once your fix is in.

**Comment by jmilloy** [comment]
Re: out of date

The time coordinates in the requested coordinates are no longer valid. You can just remove the ctype: 'midpoint' property, and the time will coordinates will default to ctype: 'point'. If you regenerate the json, I think it should leave out the ctype properties.

This is a confluence of a couple of changes.
 - I added segment_lengths. When a segment ctype, such as 'midpoint', is used, the segment lengths can be inferred in some cases (uniform coordinates and ordered numerical arrays) but must be specified in other cases (unordered numerical lists, all datetime lists, and singletons).
 - The default ctypes have changed. Whenever segment lengths cannot be inferred, the ctype defaults to 'point' instead of 'midpoint'.
 - The definition used to always include the ctype and other properties, but now it only includes these properties if they were specified by the user.

**Comment by dsully1992** [comment]
@jmilloy Thank you for the fix and explanation. I'll kick off my update of the Lambda function and test it either this afternoon or tomorrow.

**Comment by dsully1992** [comment]
Uh oh. I get the following stacktrace when following the notebook after merging develop: 

```python
AttributeError                            Traceback (most recent call last)
<ipython-input-11-9b85ecc184f8> in <module>()
----> 1 downscaled_output  = downscaled_sm.eval(coordinates)

~/Documents/gitrepos/pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Documents/gitrepos/pipeline/podpac/podpac/core/algorithm/algorithm.py in eval(self, coordinates, output)
     72         inputs = {}
     73         for key, node in self._inputs.items():
---> 74             inputs[key] = node.eval(coordinates)
     75 
     76         # accumulate output coordinates

~/Documents/gitrepos/pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in eval(self, coordinates, output)
    251 
    252         outputs = self.iteroutputs(coordinates)
--> 253         output = self.composite(outputs, output)
    254         return output
    255 

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in composite(self, outputs, result)
    302         if result is None:
    303             # consume the first source output
--> 304             result = next(outputs).copy()
    305 
    306         # initialize the mask

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in iteroutputs(self, coordinates)
    227             output = None # scratch space
    228             for src in src_subset:
--> 229                 output = src.eval(coordinates, output)
    230                 yield output
    231                 #output[:] = np.nan

~/Documents/gitrepos/pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in eval(self, coordinates, output)
    251 
    252         outputs = self.iteroutputs(coordinates)
--> 253         output = self.composite(outputs, output)
    254         return output
    255 

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in composite(self, outputs, result)
    302         if result is None:
    303             # consume the first source output
--> 304             result = next(outputs).copy()
    305 
    306         # initialize the mask

~/Documents/gitrepos/pipeline/podpac/podpac/core/compositor.py in iteroutputs(self, coordinates)
    227             output = None # scratch space
    228             for src in src_subset:
--> 229                 output = src.eval(coordinates, output)
    230                 yield output
    231                 #output[:] = np.nan

~/Documents/gitrepos/pipeline/podpac/podpac/core/data/datasource.py in eval(self, coordinates, output)
    299         # TODO move this if WCS can be updated to support
    300         self._evaluated_coordinates = coordinates
--> 301         return self._eval(coordinates, output=output)
    302 
    303     @node_eval

~/Documents/gitrepos/pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Documents/gitrepos/pipeline/podpac/podpac/core/data/datasource.py in _eval(self, coordinates, output)
    307         # TODO: support coordinate_index_type parameter to define other index types
    308         self._requested_source_coordinates, self._requested_source_coordinates_index = \
--> 309             self.native_coordinates.intersect(coordinates, outer=True, return_indices=True)
    310 
    311         # If requested coordinates and native coordinates do not intersect, shortcut with nan UnitsDataArary

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/coordinates.py in intersect(self, other, outer, return_indices)
    831         """
    832 
--> 833         intersections = [c.intersect(other, outer=outer, return_indices=return_indices) for c in self.values()]
    834         if return_indices:
    835             coords = Coordinates([c for c, I in intersections])

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/coordinates.py in <listcomp>(.0)
    831         """
    832 
--> 833         intersections = [c.intersect(other, outer=outer, return_indices=return_indices) for c in self.values()]
    834         if return_indices:
    835             coords = Coordinates([c for c, I in intersections])

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/coordinates1d.py in intersect(self, other, return_indices, outer)
    401             return self._select_empty(return_indices)
    402 
--> 403         return self.select(other.bounds, return_indices=return_indices, outer=outer)
    404 
    405     def select(self, bounds, return_indices=False, outer=False):

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/array_coordinates1d.py in select(self, bounds, outer, return_indices)
    411 
    412         # full
--> 413         if self.bounds[0] >= bounds[0] and self.bounds[1] <= bounds[1]:
    414             return self._select_full(return_indices)
    415 

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/array_coordinates1d.py in bounds(self)
    316         if self.size == 0:
    317             lo, hi = np.nan, np.nan
--> 318         elif self.is_monotonic:
    319             lo, hi = sorted([self.coords[0], self.coords[-1]])
    320         elif self.dtype is np.datetime64:

~/Documents/gitrepos/pipeline/podpac/podpac/core/coordinates/array_coordinates1d.py in is_monotonic(self)
    298     @property
    299     def is_monotonic(self):
--> 300         return self._is_monotonic
    301 
    302     @property

AttributeError: 'ArrayCoordinates1d' object has no attribute '_is_monotonic'
```

**Comment by jmilloy** [comment]
Alright, I can take a look at it.

**Comment by dsully1992** [comment]
I'm no traitlets expert, but this seems like a symptom of not having default values for the `_is_monotonic`, `_is_descending`, and `_is_uniform` properties.

**Comment by jmilloy** [comment]
I don't want a default value, but I need to figure out how an ArrayCoordinates1d object got created without setting _is_monotonic in __init__

**Comment by dsully1992** [comment]
Ok. For what it's worth, here are the steps I used to create this problem: 

```python
import podpac
smap = podpac.datalib.smap.SMAP(interpolation='bilinear')
smap.get_shared_coordinates() # or smap.shared_coordinates
```

**Comment by dsully1992** [comment]
I needed to clear my cache.

**Comment by dsully1992** [comment]
Ok. The Lambda function is updated, and works for the example in the [notebook](https://github.com/creare-com/podpac-examples/blob/develop/notebooks/demos/SMAP-downscaling-example-application.ipynb). I have an issue with loading that `pkl` file in the podpac `aws_lambda` node, which I'm working on addressing now.

**Comment by dsully1992** [comment]
Loading the pickle file was failing because my python environment had an old version (0.23.x), which I upgraded (to 0.24.1). I don't see pandas listed in `setup.py` (or `aws_requirements.txt`), and I'm not sure the best way to install it. I think it comes from xarray as it stands.

I also though `setup.py` was pulling it's required dependencies from `aws_requirements.txt`, but it doesn't appear that is the case. That is also concerning, and will likely result in the Lambda function missing dependences on a future update.

With that, I am closing this issue. We can continue discussing how to ensure the correct version of pandas is installed to load pickle files generated by the python 3.7 environment running in the Lambda function. I'm not sure if that will differ for python 2 environments.
